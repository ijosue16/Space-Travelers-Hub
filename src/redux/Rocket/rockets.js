import { createAction, createAsyncThunk, createReducer } from '@reduxjs/toolkit';

const GET_ROCKETS = 'space traveler hub/rockets/GET_ROCKETS';
const RESERVE_ROCKETS = 'space traveler hub/rockets/RESERVE_ROCKETS';
const CANCEL_RESERVATION = 'space traveler hub/rockets/CANCEL_RESERVATION';
const URL = 'https://api.spacexdata.com/v3/rockets';
const INITIALSTATE = [];

const getRocketsksAPI = async () => {
  const resp = await fetch(URL);
  const data = await resp.json();
  return data;
};
export const getRockets = createAsyncThunk(GET_ROCKETS, async (e, thunkAPI) => {
  const currentState = thunkAPI.getState();
  if (currentState.Rocketz.length === 0) {
    const resp = await getRocketsksAPI();
    const rocketsData = [];
    resp.forEach((rocket) => {
      const singleRocket = {
        id: rocket.id,
        name: rocket.rocket_name,
        description: rocket.description,
        image: rocket.flickr_images[0],
        reserved: false,
      };
      rocketsData.push(singleRocket);
    });
    return rocketsData;
  }
  return currentState.Rocketz;
});

export const reserveRocket = createAction(RESERVE_ROCKETS, (id) => ({
  payload: id,
}));

export const cancelRocketReservation = createAction(CANCEL_RESERVATION, (id) => ({
  payload: id,
}));

const Rocketreducer = createReducer(INITIALSTATE, ((builder) => {
  builder
    .addCase(getRockets.fulfilled, ((state, action) => action.payload))

    .addCase(reserveRocket, ((state, action) => {
      const newState = state.map((rocket) => {
        if (rocket.id !== action.payload) {
          return rocket;
        }
        return { ...rocket, reserved: true };
      });
      return newState;
    }))
    .addCase(cancelRocketReservation, ((state, action) => {
      const newState = state.map((rocket) => {
        if (rocket.id !== action.payload) {
          return rocket;
        }
        return { ...rocket, reserved: false };
      });
      return newState;
    }))
    .addDefaultCase();
}));

export default Rocketreducer;
