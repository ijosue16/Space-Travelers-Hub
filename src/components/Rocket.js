import React from 'react';
import 'bootstrap/dist/css/bootstrap.min.css';
import PropTypes from 'prop-types';
// /////// changes added
import { useDispatch } from 'react-redux';
import { cancelRocketReservation, reserveRocket } from '../redux/Rocket/rockets';

const Rocket = (props) => {
  const {
    id, name, description, image, reserved,
  } = props;

  // /////// changes added
  const dispatch = useDispatch();
  // const handleButtonclick = () => {
  //   console.log(id);
  // };

  // /////// changes added
  const ReserveRockethandler = () => {
    dispatch(reserveRocket(id));
    console.log(id);
  };

  const RockeReserveCancelthandler = () => {
    dispatch(cancelRocketReservation(id));
    console.log(id);
  };
  return (
    <>

      <div className="card mb-3 border-0 rounde-0 ">
        <div className="row g-0">
          <div className="col-md-3">
            <img src={image} className="bd-placeholder-img img-fluid " width="100%" height="10" alt={name} />
          </div>
          <div className="col-md-9">
            <div className="card-body pt-md-0">
              <h5 className="card-title">
                {name}
                {' '}
              </h5>
              {/* ////////////chnges added */}
              <p className="card-text">
                {/* {description} */}
                { reserved ? [<span key={0}><span className="badge bg-success py-2" key={0}>Reserved</span></span>, '   ', description] : description }
              </p>
              {/* ////////////chnges added */}
              { reserved
                ? (
                  <button
                    type="button"
                    className="btn btn-outline-secondary"
                    onClick={RockeReserveCancelthandler}
                  >
                    Cancel Reservation
                  </button>
                )
                : (
                  <button
                    type="button"
                    className="btn btn-primary "
                    onClick={ReserveRockethandler}
                  >
                    Reserve rocket
                  </button>
                )}
              {/* <button type="button" className="btn  btn-primary "
               onClick={ReserveRockethandler(id)}>Reserve rocket</button>
              <button type="button"
              class="btn btn-outline-secondary">Cancel Reservation</button> */}
            </div>
          </div>
        </div>
      </div>

    </>
  );
};
Rocket.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
};
// Rocket.defaultProps = {
//   reserved: false,
// };
export default Rocket;
