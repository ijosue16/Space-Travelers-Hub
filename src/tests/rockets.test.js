import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import Rocket from '../components/Rocket';
import store from '../redux/configurestore';
import '@testing-library/jest-dom';

describe('rockets test', () => {
  test('check if rocket renders empty initially', () => {
    render(<Provider store={store}><Rocket /></Provider>);
    expect(screen.queryByText('falcon')).not.toBeInTheDocument();
  });

  test('Rockets renders correctlly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Rocket />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
