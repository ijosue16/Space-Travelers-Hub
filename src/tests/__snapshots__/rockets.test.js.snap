// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rockets test Rockets renders correctlly 1`] = `
<div
  className="card mb-3 border-0 rounded-0 "
>
  <div
    className="row g-0"
  >
    <div
      className="col-md-3"
    >
      <img
        className="bd-placeholder-img img-fluid "
        height="10"
        width="100%"
      />
    </div>
    <div
      className="col-md-9"
    >
      <div
        className="card-body pt-md-0"
      >
        <h5
          className="card-title"
        >
           
        </h5>
        <p
          className="card-text"
        />
        <button
          className="btn btn-primary "
          onClick={[Function]}
          type="button"
        >
          Reserve rocket
        </button>
      </div>
    </div>
  </div>
</div>
`;
